### PR TITLE
Fix prematurely ending server span on undertow

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2499

Currently we end server span on undertow from `ExchangeCompletionListener` which is called when response is sent to the client. This may happen before request processing code has completed. Change span ending logic so that span is ended when all handlers have exited and exchange has been completed.
@pavolloffay could you verify that this fixes the issue you had

### CODE REVIEW

**Comment by pavolloffay** [comment]
thanks for the fix @laurit. I will have a look at this shortly.

**Comment by pavolloffay** [comment]
Maybe there could be a test that adds a filter and verifies that span is recording events after the `.doFilter`

File: instrumentation/undertow/javaagent/undertow-javaagent.gradle [hunk 1]
```diff
@@ -4,7 +4,8 @@ muzzle {
   pass {
     group = "io.undertow"
     module = 'undertow-core'
-    versions = "[2.0.0.Final,)"
+    versions = "[1.4.0.Final,)"
```
Comment (iNikem) [inline instrumentation/undertow/javaagent/undertow-javaagent.gradle:7]: Change README as well, please

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/undertow/UndertowRequestContext.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package io.opentelemetry.instrumentation.api.undertow;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.ContextKey;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/** Helper container for keeping track of request processing state in undertow. */
+public class UndertowRequestContext {
```
Comment (mateuszrzeszutek) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/undertow/UndertowRequestContext.java:13]: Can you move it to `javaagent-api`? This class does not need to be available in the library instrumentations.

Also, nit: please make it final and add a private constructor.

File: instrumentation/undertow/javaagent/undertow-javaagent.gradle (hunk 1 above)
Comment (laurit) [reply instrumentation/undertow/javaagent/undertow-javaagent.gradle:7]: added undertow to supported-libraries.md

File: instrumentation/undertow/javaagent/src/test/groovy/UndertowServerTest.groovy [hunk 3]
```diff
@@ -66,4 +93,65 @@ class UndertowServerTest extends HttpServerTest<Undertow> implements AgentTestTr
...
+          hasNoParent()
+          name "HTTP GET"
+
+          event(0) {
+            eventName "before-event"
+          }
+          event(1) {
+            eventName "after-event"
+          }
+          errorEvent(Exception, "exception after sending response", 2)
```
Comment (trask) [inline instrumentation/undertow/javaagent/src/test/groovy/UndertowServerTest.groovy:151]: can you add `attributes` validation also? e.g. would be good to see that we're capturing `200` even though also capturing the exception

**Review by trask — APPROVED** [review]
thx for tracking this down!

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**